### PR TITLE
Add callback to modify/ update logs generated by enclaves

### PR DESCRIPTION
**Pull request by RRathna**

### CODE REVIEW

**Comment by RRathna** [comment]
bors try

**Comment by RRathna** [comment]
bors try

File: enclave/core/tracee.c [hunk 1]
```diff
@@ -69,27 +77,75 @@ void oe_log_init_ecall(const char* enclave_path, uint32_t log_level)
...
+            (oe_enclave_log_callback)(
+                oe_enclave_log_context,
+                level,
+                (uint64_t)oe_thread_self(),
+                message);
+        }
+        else
+        {
+            // Skip logging for non-debug-allowed enclaves
+            if (!is_enclave_debug_allowed())
```
Comment (achamayou) [inline enclave/core/tracee.c:118]: Unless there is a specific use case to the contrary, I would suggest that the `is_enclave_debug_allowed()` check be lifted out to 103, where it applies regardless and outside the lock.

Without a clarification helping a user navigate how to sanitise log entries, trying to handle correctly them in a non-debug context seems risky at best?

File: include/openenclave/tracee.h [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+#include <openenclave/log.h>
+
+OE_EXTERNC_BEGIN
+
+typedef void (*oe_enclave_log_callback_t)(
+    void* context,
+    oe_log_level_t level,
+    uint64_t host_thread_id,
+    const char* message);
+oe_result_t oe_enclave_log_set_callback(
```
Comment (achamayou) [inline include/openenclave/tracee.h:18]: A docstring would be good here, mentioning that the callback will always be invoked under the log lock for example is useful to know.

**Comment by achamayou** [comment]
@RRathna thank you for adding this very useful feature!

**Comment by RRathna** [comment]
bors try

File: enclave/core/tracee.c (hunk 1 above)
Comment (RRathna) [reply enclave/core/tracee.c]: Thank you for pointing this out! It is better to not have the user handle the enclave logs outside of debug context. Moved it back as part of basic checks, irrespective callback.

File: include/openenclave/tracee.h (hunk 2 above)
Comment (RRathna) [reply include/openenclave/tracee.h:52]: Added docstring.

File: include/openenclave/tracee.h [hunk 3]
```diff
@@ -0,0 +1,58 @@
+// Copyright (c) Open Enclave SDK contributors.
+// Licensed under the MIT License.
+
+#ifndef _OE_TRACE_E_H
+#define _OE_TRACE_E_H
```
Comment (dthaler) [inline include/openenclave/tracee.h:5]:
```suggestion
#ifndef _OE_TRACEE_H
#define _OE_TRACEE_H
```
Match filename

File: include/openenclave/tracee.h [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+/**
+ * Define callback to process logs generated by enclave.
+ *
+ * The callback, defined as expected, will be called every time OE_TRACE is
+ * is called by enclave routines, irrespective of log level. The callback will
+ * be processed under log lock.
+ *
+ * NOTE: The function oe_enclave_log_set_callback sets the callback and only
+ * then enclave logs will be processed through the callback function.
```
Comment (dthaler) [inline include/openenclave/tracee.h:21]:
```suggestion
 * then will enclave logs be processed through the callback function.
```
grammar nit

File: include/openenclave/tracee.h [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+ *
+ * @param[in] context The parameter can be used by the user to send in
+ * additional variables/ structures that might be needed to process the logs.
```
Comment (dthaler) [inline include/openenclave/tracee.h:24]:
```suggestion
 * an additional variable or structure that might be needed to process the logs.
```

File: include/openenclave/tracee.h [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+ * @param[in] message The parameter holds the enclave log message generated.
+ */
+typedef void (*oe_enclave_log_callback_t)(
+    void* context,
+    oe_log_level_t level,
+    uint64_t thread_id,
+    const char* message);
+
+/**
+ * Set callback that will process logs generated by enclave.
```
Comment (dthaler) [inline include/openenclave/tracee.h:38]:
```suggestion
 * Set callback that will process logs generated by the enclave.
```

File: include/openenclave/tracee.h [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+ *
+ * This function, when called ensures the enclave logs are transfered to
```
Comment (dthaler) [inline include/openenclave/tracee.h:40]:
```suggestion
 * This function, when called ensures the enclave logs are transferred to
```

File: include/openenclave/tracee.h [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+ * the preferred callback instead of the regular route.
+ *
+ * @param[in] context The parameter holds a pointer to variable/
```
Comment (dthaler) [inline include/openenclave/tracee.h:43]:
```suggestion
 * @param[in] context A pointer to a variable or
```

File: include/openenclave/tracee.h [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+ * structure that can be used to process the logs inside the callback
+ * function.
+ * @param[in] callback The parameter holds pointer to function that is
```
Comment (dthaler) [inline include/openenclave/tracee.h:46]:
```suggestion
 * @param[in] callback A pointer to a function that will
```

File: include/openenclave/tracee.h [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+ * be assigned as the callback to process enclave logs.
+ *
+ * @returns OE_OK successful
+ * @returns OE_FAILURE failed to add callback
```
Comment (dthaler) [inline include/openenclave/tracee.h:50]:
```suggestion
 * @retval OE_OK The operation was successful.
 * @retval OE_FAILURE Failed to add callback.
```
Use retval for discrete values.

File: samples/log_callback/enclave/enc.c [hunk 11]
```diff
@@ -1,13 +1,53 @@
...
+        modified_log,
+        "E, %s, %llx, %s",
+        oe_log_level_strings[level],
+        thread_id,
+        message);
+
+    /*
+     * Add logic here to modify the log message, to obscure enclave logs from
+     * the host. The context can be used to transfer shared/public/private
+     * secret, that maybe used to encrypt the log message.
```
Comment (dthaler) [inline samples/log_callback/enclave/enc.c:31]:
```suggestion
     * the host. The context might be used, for example, to transfer a shared/public/private
     * secret that may be used to encrypt the log message.
```

File: samples/log_callback/enclave/enc.c [hunk 12]
```diff
@@ -1,13 +1,53 @@
...
+     *
+     * NOTE: fprintf does not work consitently in this function. If used, the
```
Comment (dthaler) [inline samples/log_callback/enclave/enc.c:33]:
```suggestion
     * NOTE: fprintf does not work consistently in this function. If used, the
```

Comment (dthaler) [reply samples/log_callback/enclave/enc.c:33]: Is there a bug?  Is there an issue filed?

File: samples/log_callback/log_callback.edl [hunk 13]
```diff
@@ -4,12 +4,15 @@
 enclave {
     from "openenclave/edl/syscall.edl" import *;
     from "platform.edl" import *;
+    from "openenclave/edl/logging.edl" import *;
```
Comment (dthaler) [inline samples/log_callback/log_callback.edl:7]:
```suggestion
    from "openenclave/edl/logging.edl" import *;
    from "platform.edl" import *;
```

Use alphabetical order whenever possible.

File: samples/log_callback/enclave/enc.c (hunk 12 above)
Comment (RRathna) [reply samples/log_callback/enclave/enc.c:33]: Yes, this is a bug, ( might not be solvable). This bug will be hit only when fprintf (using oehostfs) is called after oe_terminate_enclave() call. I guess logging is the only place where this could happen. I wanted to create an issue and modify the NOTE in code after the sample gets checked-in.

**Comment by RRathna** [comment]
bors try

File: include/openenclave/tracee.h [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+ * @param[in] context A pointer to variable or structure that can be used to
+ * process the logs inside the callback function.
+ * @param[in] callback The parameter holds pointer to function that is be
+ * assigned as the callback to process enclave logs.
+ *
+ * @retval OE_OK Callback was added successfully.
+ * @retval OE_FAILURE Failed to add callback.
+ */
+oe_result_t oe_enclave_log_set_callback(
+    void* context,
```
Comment (mingweishih) [inline include/openenclave/tracee.h:52]: From the API perspective, I would suggest add the `context_size` variable so that the function don't blindly take an opaque pointer.

File: enclave/core/tracee.c [hunk 15]
```diff
@@ -69,6 +77,62 @@ void oe_log_init_ecall(const char* enclave_path, uint32_t log_level)
...
+            (oe_enclave_log_callback)(
+                oe_enclave_log_context,
+                level,
+                (uint64_t)oe_thread_self(),
+                message);
+        }
+        else
+        {
+            // Check if this message should be skipped
+            if (level > _active_log_level)
```
[the remark this answers is not in the document]
Comment (RRathna) [reply enclave/core/tracee.c:118]: Since log_level is being passed directly to callback, the logic inside the callback can decide to filter out or not.

File: include/openenclave/tracee.h [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+ * be processed under log lock.
+ *
+ * NOTE: The function oe_enclave_log_set_callback sets the callback and only
+ * then will enclave logs be processed through the callback function.
+ *
+ * @param[in] context The parameter can be used by the user to send in an
+ * additional variable or structures that might be needed to process the logs.
+ * @param[in] level The parameter holds the log level of the current enclave
+ * log being processed and can be used by the callback.
+ * @param[in] thread_id The parameter holds the thread id of the curent
```
[the remark this answers is not in the document]
Comment (RRathna) [reply include/openenclave/tracee.h:27]: The requirement was that the callback have the same data as the host side callback (minus time). I adopted the same format as https://github.com/openenclave/openenclave/blob/6ef7c19fe3ec9e9762c91b8edbaf5bd170d2fb58/host/traceh.c#L376

File: include/openenclave/tracee.h (hunk 14 above)
Comment (RRathna) [reply include/openenclave/tracee.h:52]: I like this suggestion. Would be safer to add size of the structure, if any.

@achamayou, do you think its okay to deviate from host callback API and have this additional parameter in the set_callback API?

Comment (achamayou) [reply include/openenclave/tracee.h:52]: @RRathna sure, that's fine. Ultimately the implementation needs to know what to cast to, but having the size does provide some safety.

File: samples/log_callback/enclave/enc.c (hunk 12 above)
Comment (dthaler) [reply samples/log_callback/enclave/enc.c]: Is there an issue filed?  Would be nice to add the issue number into the comment. Alternatively if it is by design (not a bug) then expand comment to explain the issue, like explaining that fprintf intentionally will not work after the oe_terminate_enclave() call.

File: include/openenclave/tracee.h (hunk 14 above)
Comment (dthaler) [reply include/openenclave/tracee.h:52]: Advice from me as the API maintainer: If the context is always being passed back to the same code that set the context then you do not need a context size.  If the caller cares about the size, it can encode it in the first field of the context structure.   Unless more than one piece of code (i.e., from different developers) needs it, one should not add a context size.

File: enclave/core/tracee.c [hunk 17]
```diff
@@ -69,6 +77,66 @@ void oe_log_init_ecall(const char* enclave_path, uint32_t log_level)
...
+
+static oe_result_t oe_log_enclave_message(
+    oe_log_level_t level,
+    const char* message)
+{
+    oe_result_t result = OE_FAILURE;
+
+    if (!message)
+    {
+        result = OE_UNEXPECTED;
```
Comment (mingweishih) [inline enclave/core/tracee.c:109]: Simply do `OE_RAISE(OE_INVALID_PARAMETER)`

File: samples/log_callback/enclave/enc.c [hunk 18]
```diff
@@ -1,13 +1,54 @@
...
+        thread_id,
+        message);
+
+    /*
+     * Add logic here to modify the log message, to obscure enclave logs from
+     * the host. The context might be used, for example, to transfer a
+     * shared/public/private secret, that may be used to encrypt the log
+     * message.
+     *
+     * NOTE: fprintf does not work consistently in this function. If used, the
```
Comment (mingweishih) [inline samples/log_callback/enclave/enc.c:34]: Please open an issue for this bug. I think this can be rephrased as all operations based on OE's host filesystem support would fail.

File: include/openenclave/tracee.h (hunk 14 above)
Comment (mingweishih) [reply include/openenclave/tracee.h:52]: I'm fine with Dave's suggestion. I guess it's just unclear to me how the context could be used. It would be helpful to demonstrate some use cases in the sample as well (instead of passing NULL in the current sample).

File: enclave/core/tracee.c [hunk 19]
```diff
@@ -69,6 +77,66 @@ void oe_log_init_ecall(const char* enclave_path, uint32_t log_level)
...
+        goto done;
+    }
+    // Take the log file lock.
```
Comment (mingweishih) [inline enclave/core/tracee.c:112]: Nit: add empty line above

**Review by mingweishih — APPROVED** [review]
Looks good to me. Approved with expectations of addressing pending comments.

File: include/openenclave/tracee.h (hunk 14 above)
Comment (achamayou) [reply include/openenclave/tracee.h:52]: In cases where the user app wants to log to something other than a file, this may point to a network context for example, or a message queue.

File: include/openenclave/tracee.h (hunk 16 above)
[the remark this answers is not in the document]
Comment (mingweishih) [reply include/openenclave/tracee.h:27]: Please update the comment, something like
```suggestion
 * @param[in] thread_id The parameter holds the thread id (the base address of thread data) of the current
```

File: include/openenclave/tracee.h [hunk 20]
```diff
@@ -0,0 +1,57 @@
...
+#include <openenclave/log.h>
+
+OE_EXTERNC_BEGIN
+
+/**
+ * User-defined callback for processing logs generated by the enclave.
+ *
+ * The callback, defined as expected, will be called every time OE_TRACE is
+ * is called by enclave routines, irrespective of log level. The callback
+ * will be processed under log lock.
```
Comment (dthaler) [inline include/openenclave/tracee.h:18]:
```suggestion
 * will be processed under the log lock.
```
grammar nit

File: include/openenclave/tracee.h [hunk 21]
```diff
@@ -0,0 +1,57 @@
...
+ *
+ * NOTE: The function oe_enclave_log_set_callback sets the callback and only
+ * then will enclave logs be processed through the callback function.
+ *
+ * @param[in] context The parameter can be used by the user to send in an
+ * additional variable or structures that might be needed to process the logs.
```
Comment (dthaler) [inline include/openenclave/tracee.h:24]:
```suggestion
 * additional variable or structure that might be needed to process the logs.
```

File: include/openenclave/tracee.h [hunk 22]
```diff
@@ -0,0 +1,57 @@
...
+ * @param[in] level The parameter holds the log level of the current enclave
+ * log being processed and can be used by the callback.
+ * @param[in] thread_id The parameter holds the thread id of the curent
```
Comment (dthaler) [inline include/openenclave/tracee.h:27]:
```suggestion
 * @param[in] thread_id The parameter holds the thread id of the current
```

File: include/openenclave/tracee.h [hunk 23]
```diff
@@ -0,0 +1,57 @@
...
+    uint64_t thread_id,
+    const char* message);
+
+/**
+ * Set callback that will process logs generated by the enclave.
+ *
+ * This function, when called ensures the enclave logs are transferred to
+ * the preferred callback instead of the regular route.
+ *
+ * @param[in] context A pointer to variable or structure that can be used to
```
Comment (dthaler) [inline include/openenclave/tracee.h:43]:
```suggestion
 * @param[in] context A pointer to a variable or structure that can be used to
```

File: include/openenclave/tracee.h [hunk 24]
```diff
@@ -0,0 +1,57 @@
...
+ * process the logs inside the callback function.
+ * @param[in] callback The parameter holds pointer to function that is be
```
Comment (dthaler) [inline include/openenclave/tracee.h:45]:
```suggestion
 * @param[in] callback Pointer to a function that is be
```

File: include/openenclave/tracee.h (hunk 14 above)
Comment (dthaler) [reply include/openenclave/tracee.h:52]: Example: app calls the set callback function and passes in as context, a pointer to a global struct.   In the callback, app casts the context pointer back to its own struct pointer.

**Review by dthaler — APPROVED** [review]
Ok except for grammar (see suggested changes) and the need to file an issue and add issue # into a code comment.

File: include/openenclave/tracee.h (hunk 14 above)
Comment (RRathna) [reply include/openenclave/tracee.h:52]: Will add sample with context set to a global structure in a seperate PR.

**Comment by radhikaj** [comment]
bors try

**Comment by RRathna** [comment]
bors try

File: samples/log_callback/enclave/enc.c (hunk 12 above)
Comment (RRathna) [reply samples/log_callback/enclave/enc.c]: Added issue #4349 to comments in enc.c:enclave_customized_log.

File: enclave/core/tracee.c [hunk 25]
```diff
@@ -69,6 +77,64 @@ void oe_log_init_ecall(const char* enclave_path, uint32_t log_level)
...
+        {
+            result = OE_OK;
+            goto done;
+        }
+
+        OE_CHECK(oe_log_ocall(level, message));
+        goto done;
+    }
+
+    oe_mutex_unlock(&_log_lock);
```
Comment (radhikaj) [inline enclave/core/tracee.c:131]: On line 127, we could go to done and miss the unlock?

File: include/openenclave/tracee.h [hunk 26]
```diff
@@ -0,0 +1,57 @@
+// Copyright (c) Open Enclave SDK contributors.
+// Licensed under the MIT License.
+
+#ifndef _OE_TRACEE_H
+#define _OE_TRACEE_H
+
+#include <openenclave/bits/defs.h>
+#include <openenclave/enclave.h>
+#include <openenclave/log.h>
```
Comment (radhikaj) [inline include/openenclave/tracee.h:9]: Do you need to include log.h here?

Comment (RRathna) [reply include/openenclave/tracee.h:9]: log.h is needed for "oe_log_level_t" used in the callback function pointer declaration

**Comment by RRathna** [comment]
bors try

File: enclave/core/tracee.c [hunk 27]
```diff
@@ -69,6 +77,63 @@ void oe_log_init_ecall(const char* enclave_path, uint32_t log_level)
...
+            goto done;
+        }
+
+        OE_CHECK_NO_TRACE(oe_log_ocall(level, message));
+    }
+
+    result = OE_OK;
+
+done:
+    oe_mutex_unlock(&_log_lock);
```
Comment (radhikaj) [inline enclave/core/tracee.c:133]: . If !message you will go to done without locking the mutex. can you verify that if the mutex is not locked, that this does not cause any issues?

File: enclave/core/tracee.c [hunk 28]
```diff
@@ -84,13 +149,6 @@ oe_result_t oe_log(oe_log_level_t level, const char* fmt, ...)
         goto done;
     }
 
-    // Check if this message should be skipped
```
Comment (radhikaj) [inline enclave/core/tracee.c:87]: why is this deleted?

File: samples/log_callback/host/host.c [hunk 29]
```diff
@@ -45,7 +45,7 @@ void customized_log(
     strftime(time, sizeof(time), "%Y-%m-%dT%H:%M:%S%z", t);
 
     FILE* log_file = NULL;
-    if (level >= OE_LOG_LEVEL_WARNING)
+    if (level >= OE_LOG_LEVEL_ERROR)
```
Comment (radhikaj) [inline samples/log_callback/host/host.c:48]: Is there a reason to change this?

File: samples/log_callback/host/host.c [hunk 30]
```diff
@@ -65,10 +65,16 @@ void customized_log(
         message);
 }
 
+void host_transfer_logs_to_file(const char* modified_log, size_t size)
+{
+    FILE* enc_logfile = fopen("./oe_enclave_out.txt", "w");
+    fprintf(enc_logfile, "%.*s", (int)size, modified_log);
```
Comment (radhikaj) [inline samples/log_callback/host/host.c:71]: fclose after this.

File: enclave/core/tracee.c [hunk 31]
```diff
@@ -13,13 +12,22 @@
...
 #include <openenclave/internal/utils.h>
+#include <openenclave/tracee.h>
+
 #include "core_t.h"
+#include "tracee.h"
 
 static oe_log_level_t _active_log_level = OE_LOG_LEVEL_ERROR;
 static char _enclave_filename[OE_MAX_FILENAME_LEN];
 
+const char* const oe_log_level_strings[OE_LOG_LEVEL_MAX] =
```
Comment (anakrish) [inline enclave/core/tracee.c:26]: Can this be made static like the variables above? Symbols that are not static are part of OE's binary interface which ought not be expanded unless needed.

File: enclave/core/tracee.c [hunk 32]
```diff
@@ -69,6 +77,63 @@ void oe_log_init_ecall(const char* enclave_path, uint32_t log_level)
...
+{
+    oe_result_t result = OE_OK;
+
+    OE_CHECK_NO_TRACE(oe_mutex_lock(&_log_lock));
+
+    oe_enclave_log_context = context;
+    oe_enclave_log_callback = callback;
+    oe_mutex_unlock(&_log_lock);
+
+    OE_TRACE_INFO("enclave logging callback is set");
```
Comment (anakrish) [inline enclave/core/tracee.c:95]: Is this trace safe given that it is part of logging/tracing itself.
If yes, add a comment that it does not introduce a recursive call.

File: enclave/core/tracee.c [hunk 33]
```diff
@@ -69,6 +77,63 @@ void oe_log_init_ecall(const char* enclave_path, uint32_t log_level)
     }
 }
 
+static void* oe_enclave_log_context = NULL;
```
Comment (anakrish) [inline enclave/core/tracee.c:80]: Naming convention. static variables don't have `oe_` prefix; see `_log_lock` above.

File: enclave/core/tracee.c (hunk 28 above)
Comment (RRathna) [reply enclave/core/tracee.c:87]: This is moved into oe_log_enclave_message(). This check is relevant only when callback is not set.

File: enclave/core/tracee.c (hunk 31 above)
Comment (RRathna) [reply enclave/core/tracee.c:26]: This can't be made static because we are using the variable that was defined as an extern in https://github.com/openenclave/openenclave/blob/master/include/openenclave/log.h#L20.

I followed the example of host side logging functions - 
https://github.com/openenclave/openenclave/blob/master/host/traceh.c#L32

File: enclave/core/tracee.c (hunk 32 above)
Comment (RRathna) [reply enclave/core/tracee.c:95]: since we call this after the basic logging functions are already setup, as a function call from the enclave application itself, this trace is safe.. I'll add a comment to the source.

File: samples/log_callback/host/host.c (hunk 30 above)
Comment (RRathna) [reply samples/log_callback/host/host.c:76]: Added fclose for enc_logfile as part of label exit: in main().

**Review by anakrish — APPROVED** [review]

**Comment by RRathna** [comment]
bors try

**Comment by RRathna** [comment]
bors try

**Comment by RRathna** [comment]
bors try

**Comment by radhikaj** [comment]
bors try

**Comment by mingweishih** [comment]
bors try

**Comment by radhikaj** [comment]
Bors r+
